netconf_client/ncclient.py

```diff
@@ -21,40 +24,258 @@
         :type session: :class:`netconf_client.session.Session`
 
         :param float timeout: Duration in seconds to wait for replies
+        :param string log_id: log ID string printed with each log entry,
+               the peer's IP address as default
         """
         self.timeout = timeout
         self.session = session
+        self.log_id = log_id
+        self.start_time = self._get_timestamp()
+        self.log_level = logging.DEBUG
+
+        self.logged_functions = {
+            "get": True,
+            "edit_config": True,
+            "get_config": True,
+            "copy_config": True,
+            "discard_changes": True,
+            "commit": True,
+            "lock": True,
+            "unlock": True,
+            "kill_session": True,
+            "create_subscription": True,
+            "validate": True,
+            "delete_config": True,
+            "dispatch": True,
+            "close_session": False,
+        }
 
     def __enter__(self):
         return self
 
     def __exit__(self, a, b, c):
         self.session.__exit__(a, b, c)
 
+    @staticmethod
+    def get_logger():
+        """Returns the internally used logger instance (same for all sessions)"""
+        return logger
+
+    def get_log_level(self):
+        """Returns current logging level, default is DEBUG"""
+        return self.log_level
+
+    def set_log_level(self, level=logging.DEBUG):
+        """Sets the log level to use for logging"""
+        self.log_level = level
+
+    def get_logged_functions(self):
+        """Returns a dict 'str'->'Boolean' of function names to be logged"""
+        return self.logged_functions
+
+    def is_function_logged(self, funcname):
+        """Checks whether a given function is enabled for logging"""
+        return self.logged_functions.get(funcname, False)
+
+    def enable_logging(self, funcname):
+        """Enables logging for given function name"""
+        if funcname in self.logged_functions:
+            self.logged_functions.update({funcname: True})
+
+    def disable_logging(self, funcname):
+        """Disables logging for given function name"""
+        if funcname in self.logged_functions:
+            self.logged_functions.update({funcname: False})
+
+    def _get_timestamp(self):
+        return datetime.now()
+
+    def _is_logger_enabled(self, funcname):
+        return self.is_function_logged(funcname) and Manager.get_logger().isEnabledFor(
+            self.get_log_level()
+        )
+
+    def _get_peer_ip(self):
+        """Returns the peer's IP address.
+           If not connected, an empty string is returned.
+        """
+
+        addr = ""
+        try:
+            s = self.session.sock.sock
+            (addr, _) = s.getpeername()
+        except:
+            pass
+        return addr
+
+    def _get_request_log_id(self):
+        """Returns string containing device type, device num (if any),
+           and IP address
+        """
+        peer_ip = self._get_peer_ip()
```

Comment:
It may also be prudent to get the local and remote port. This could help users disambiguate multiple sessions if they didn't provide a `log_id`.

Reply:
Good point, I add this.